dqvm/src/addresses.lisp

```diff
@@ -0,0 +1,282 @@
+;;;; src/addresses.lisp
+;;;;
+;;;; Author: Juan M. Bello-Rivas
+
+(in-package #:dqvm2)
+
+;;; This file implements an address table for the amplitudes in the wave
+;;; function. The table does not explicitly contain the addresses, instead it
+;;; computes them on the fly by keeping track of the current qubit
+;;; permutation. The rationale for this approach is so that every MPI rank
+;;; can determine which addresses are held by any other MPI rank with a
+;;; minimal memory footprint.
+;;;
+;;; To illustrate the mechanism used to manage addresses in local and remote
+;;; ranks, we consider the following example program:
+;;;
+;;;     X 1
+;;;     Z 2
+;;;
+;;; There are 3 qubits (0, 1, and 2) and 2^3 = 8 addresses and amplitudes
+;;; involved in the above program. If we have 2 ranks available, then each
+;;; one of them keeps 2³ / 2 = 4 address/amplitude pairs and the the data is
+;;; split among them as shown below:
+;;;
+;;;       rank #0 | rank #1
+;;;       --------+--------
+;;;       0 1 2 3 | 4 5 6 7
+;;;
+;;; Suppose we have already executed "X 1" and we are about to execute "Z
+;;; 2". Moreover, imagine we are sitting at rank #0, then we must answer the
+;;; following questions:
+;;;
+;;;   1. Which address/amplitude pairs are going to be sent and where are
+;;;   they going (i.e., which rank and which place within the receiving
+;;;   rank)?
+;;;
+;;;   2. Which addresses and amplitudes is rank #0 going to be receiving, and
+;;;   from which source?
+;;;
+;;; To answer the questions above, it is convenient to keep the following
+;;; diagram in mind:
+;;;
+;;;       rank #0 | rank #1
+;;;       --------+--------
+;;;   π₀  0 1 2 3 | 4 5 6 7
+;;;   π₁  0 2 1 3 | 4 6 5 7  <--- Addresses after executing "X 1"
+;;;   π₂  0 4 2 6 | 1 5 3 7  <--- Addresses after executing "Z 2"
+;;;
+;;; where πₖ is the permutation associated with the k-th instruction and π₀
+;;; is the identity.
```

Comment:
Why should executing the instruction `X 1` result in `0 2 1 3...`? I naively was expecting `2 3 0 1 6 7 4 5`  (i.e. `a |000> + ... -> ... + a |010> + ...`, etc etc).

Indeed, if I `run-dqvm`

```
(dqvm2-tests::run-dqvm 3 (quil:parse-quil "X 1"))
#(#C(2.0d0 0.0d0) #C(3.0d0 0.0d0) #C(0.0d0 0.0d0) #C(1.0d0 0.0d0)
  #C(6.0d0 0.0d0) #C(7.0d0 0.0d0) #C(4.0d0 0.0d0) #C(5.0d0 0.0d0))
```

then `2 3 0 1 6 7 4 5` is the result (starting from your "debug" wavefunction with amps `0 1 2 3...`).

At the same, I realize the example here is correct because

```
(loop :for i :below 8 :collect (get-address-by-offset (addresses *dqvm*) i))
(0 2 1 3 4 6 5 7)
```

where `*dqvm*` is the one created by the above `run-dqvm` call.

Apparently I am mixed up about the difference between the πₖ which act on addresses and the action of the permutation gate `X` on the amplitudes.

Reply:
This is a good question. I have extended the comment with more information and a fully worked out example that explains your results. The gist of it is that permutations on addresses are independent of permutations induced by a gate for the purposes of DQVM.

Let me know if you still have questions after reading the extra material.

Reply:
Thanks very much for this clear explanation, and for adding it to the comments even though I was probably the only one confused about it :)

Reply:
You're welcome. I think it was totally worth it to write it down for current and future team members.